2014_TPDS_basic_decomposition_based_algorithm/src/decomposition.rs

```diff
@@ -36,6 +36,33 @@ pub fn decompose(dag: &mut Graph<NodeData, i32>) {
             (rounded_node_deadline * deadline_factor as f32) as i32,
         );
     }
+
+    // Sort because offsets need to be calculated in the order of execution.
+    let mut topological_order = Topo::new(&*dag);
+    let mut sorted_nodes: Vec<_> = Vec::new();
+    while let Some(node_i) = topological_order.next(&*dag) {
+        sorted_nodes.push(node_i);
+    }
+
+    // Add offset to nodes.
+    for node_i in sorted_nodes {
+        let pre_nodes = dag.get_pre_nodes(node_i);
+        match pre_nodes {
+            None => dag.add_param(node_i, "integer_scaled_offset", 0),
+            // offset = maximum of offset + deadline of predecessor node
+            Some(nodes) => {
+                let max_offset = nodes
+                    .iter()
+                    .map(|&pre_node_i| {
+                        dag[pre_node_i].params["integer_scaled_offset"]
+                            + dag[pre_node_i].params["integer_scaled_deadline"]
+                    })
+                    .max()
+                    .unwrap();
+                dag.add_param(node_i, "integer_scaled_offset", max_offset);
+            }
+        }
+    }
```

Comment:
```suggestion
   // Sort because offsets need to be calculated in the order of execution.
    let mut topological_order = Topo::new(&*dag);
    while let Some(node_i) = topological_order.next(&*dag) {
        match dag.get_pre_nodes(node_i) {
            None => dag.add_param(node_i, "integer_scaled_offset", 0),
            // offset = maximum of offset + deadline of predecessor node
            Some(nodes) => {
                let max_offset = nodes
                    .iter()
                    .map(|&pre_node_i| {
                        dag[pre_node_i].params["integer_scaled_offset"]
                            + dag[pre_node_i].params["integer_scaled_deadline"]
                    })
                    .max()
                    .unwrap();
                dag.add_param(node_i, "integer_scaled_offset", max_offset);
            }
        }
    }
```

Reply:
Thanks for the correction. Applied.